### PR TITLE
Profile validation for Create and Update resources

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
## Description
Add profile validator to Create and Update resources.
Trigger it only feature flag is set up, or special header been pass (header has higher priority)

## Testing
Manual and unit tests.

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Feature

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/UpsertResourceValidator.cs [hunk 1]
```diff
@@ -13,13 +16,24 @@ namespace Microsoft.Health.Fhir.Core.Features.Resources.Upsert
...
+            IOptions<CoreFeatureConfiguration> config)
         {
             RuleFor(x => x.Resource.Id)
                 .NotEmpty().WithMessage(Core.Resources.UpdateRequestsRequireId);
 
+            var contentValidator = new ResourceProfileValidator(
+              modelAttributeValidator,
+              profileValidator,
+              fhirRequestContextAccessor,
+              config.Value.ProfileValidationOnCreate);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/UpsertResourceValidator.cs:33]: Should this be `ProfileValidationOnUpdate`?

File: src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                {
+                    if (bool.TryParse(value, out bool headerValue))
+                    {
+                        profileValidation = headerValue;
+                    }
+                }
+
+                if (profileValidation)
+                {
+                    var typedElemat = resourceElement.Instance;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs:60]: nit: typo

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/UpsertResourceValidator.cs (hunk 1 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Upsert/UpsertResourceValidator.cs:33]: Yep, and this is why tests are failing.

File: src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                    var errors = _profileValidator.TryValidate(resourceElement.Instance);
+                    var fullFhirPath = resourceElement.InstanceType;
+
+                    foreach (var error in errors.Where(x => x.Severity == Severity.Error.ToString()))
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs:64]: What do other severities mean?

Comment (Ivanidzo4ka) [reply src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs:64]: https://www.netmeister.org/blog/images/doh.png
It should be Error and Fatal.
Warning is something not supported on firely side.
Error - error during validation, like cardinality is wrong
Haven't see Info and Fatal and I don't think they present at all.
https://github.com/FirelyTeam/firely-net-sdk/blob/f028de2ea5d5b937f4a51f79c1dcbe9887615204/src/Hl7.Fhir.Specification/Support/Issue.cs#L64

File: src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceContentValidator.cs [hunk 4]
```diff
@@ -3,44 +3,41 @@
...
 
 namespace Microsoft.Health.Fhir.Core.Features.Validation
 {
-    public class AttributeValidator : IPropertyValidator
+    /// <summary>
+    /// Validates content of resource.
+    /// </summary>
+    /// <remarks>
+    /// Even if we correctly parsed resource into object it doesn't mean resource is valid.
+    /// We need to check what property filled properly, they have right cardinality, correct types, proper format, etc.
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceContentValidator.cs:20]: nit: perhaps shorten to read better:
```suggestion
    /// We need to check that properties have right cardinality, correct types, proper format, etc.
```

File: src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+                if (profileValidation)
+                {
+                    var errors = _profileValidator.TryValidate(resourceElement.Instance);
+                    var fullFhirPath = resourceElement.InstanceType;
+
+                    foreach (var error in errors.Where(x => x.Severity == IssueSeverity.Error || x.Severity == IssueSeverity.Fatal))
+                    {
+                        yield return new FhirValidationFailure(
+                            fullFhirPath,
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Validation/ResourceProfileValidator.cs:67]: nit: Is this just the type name?

**Review by brendankowitz — APPROVED** [review]
🚢

**Pull request merged**